### PR TITLE
Fix Object UFS edge case again

### PR DESCRIPTION
**Pull request by bradyoo**
Adding / to the keys was not a good idea.

### CODE REVIEW

**Comment by bradyoo** [comment]
@gpang

File: underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemTest.java [hunk 1]
```diff
@@ -244,8 +244,8 @@ public void getOperationMode() throws Exception {
   public void stripPrefixIfPresent() throws Exception {
     Assert.assertEquals("", mS3UnderFileSystem.stripPrefixIfPresent("s3a://" + BUCKET_NAME));
     Assert.assertEquals("", mS3UnderFileSystem.stripPrefixIfPresent("s3a://" + BUCKET_NAME + "/"));
-    Assert.assertEquals("test/", mS3UnderFileSystem.stripPrefixIfPresent("test"));
-    Assert.assertEquals("test/",
-        mS3UnderFileSystem.stripPrefixIfPresent("s3a://" + BUCKET_NAME + "/test/"));
```
Comment (gpang) [inline underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemTest.java:249]: can we keep this test case? `"s3a://" + BUCKET_NAME + "/test/"`

Comment (bradyoo) [reply underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemTest.java:249]: No because I don't think /test/ is actually a valid input to this function. It was becuase I thought it was a valid input that I made this mistake.

**Comment by bradyoo** [comment]
alluxio-bot, test this please.

File: underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemTest.java (hunk 1 above)
Comment (gpang) [reply underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemTest.java:249]: We can't pass in directory paths to this method? I thought any s3 UFS path can be passed to this method?

Comment (gpang) [reply underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemTest.java:249]: I think directories are also allowed, since a few lines above, we have `"s3a://" + BUCKET_NAME + "/"` which is a directory.

**Comment by bradyoo** [comment]
ah right. Sorry reflected feedback. Fixed. Also, confirmed that this fixed some of the autobots issues.

File: underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemTest.java [hunk 2]
```diff
@@ -244,7 +244,7 @@ public void getOperationMode() throws Exception {
   public void stripPrefixIfPresent() throws Exception {
     Assert.assertEquals("", mS3UnderFileSystem.stripPrefixIfPresent("s3a://" + BUCKET_NAME));
     Assert.assertEquals("", mS3UnderFileSystem.stripPrefixIfPresent("s3a://" + BUCKET_NAME + "/"));
-    Assert.assertEquals("test/", mS3UnderFileSystem.stripPrefixIfPresent("test"));
+    Assert.assertEquals("test", mS3UnderFileSystem.stripPrefixIfPresent("test"));
```
Comment (gpang) [inline underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemTest.java:247]: Can we do the full matrix of test cases?

```
test
test/
/test
/test/
```

are these also possible?
```
""
"/"
```

File: core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java [hunk 3]
```diff
@@ -1118,14 +1118,25 @@ protected boolean parentExists(String path) throws IOException {
...
       return stripedKey;
     }
-    return CommonUtils.stripPrefixIfPresent(normalizedPath, PATH_SEPARATOR);
+
+    // See if adding a PATH_SEPARATOR makes a difference in stripping prefix
+    final String normalizedPath = PathUtils.normalizePath(path, PATH_SEPARATOR);
+    if (!path.equals(normalizedPath)) {
+      stripedKey = CommonUtils.stripPrefixIfPresent(
+          normalizedPath,
+          PathUtils.normalizePath(mRootKeySupplier.get(), PATH_SEPARATOR));
```
Comment (gpang) [inline core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:1133]: Can we save this normalized root at the top, so both branches can re-use it?

File: core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java [hunk 4]
```diff
@@ -1118,14 +1118,25 @@ protected boolean parentExists(String path) throws IOException {
...
-        normalizedPath,
+        path,
         PathUtils.normalizePath(mRootKeySupplier.get(), PATH_SEPARATOR));
-    if (!stripedKey.equals(normalizedPath)) {
+    if (!stripedKey.equals(path)) {
       return stripedKey;
     }
-    return CommonUtils.stripPrefixIfPresent(normalizedPath, PATH_SEPARATOR);
+
+    // See if adding a PATH_SEPARATOR makes a difference in stripping prefix
```
Comment (gpang) [inline core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:1128]: What is an example where this will make a difference? `PathUtils.normalizePath` just adds a separator at the end, so does this only apply to the single string `s3://bucket`?

File: core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java (hunk 3 above)
Comment (bf8086) [inline core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:1133]: (nit) reuse the result from line 1123?

File: core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java (hunk 4 above)
Comment (bradyoo) [reply core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:1128]: That is correct.

**Comment by bradyoo** [comment]
Added test cases and simplified code based on gene's comment.

File: core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java [hunk 5]
```diff
@@ -1118,14 +1118,16 @@ protected boolean parentExists(String path) throws IOException {
    */
   @VisibleForTesting
   public String stripPrefixIfPresent(String path) {
-    final String normalizedPath = PathUtils.normalizePath(path, PATH_SEPARATOR);
+    if (mRootKeySupplier.get().equals(path)) {
```
Comment (bradyoo) [inline core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:1121]: mRootKeySupplier.get() is a memoized function so don't be concerned about callng twice.

**Comment by bradyoo** [comment]
alluxio-bot, test this please.

**Comment by gpang** [comment]
alluxio-bot, merge this please

**Comment by gpang** [comment]
alluxio-bot, cherry-pick this to branch-2.3 please

**Comment by gpang** [comment]
alluxio-bot, cherry-pick this to branch-2.2 please
